### PR TITLE
517: Reflected Christian Gruen's remarks

### PR DESCRIPTION
**Pull request by dnovatchev**
I deleted my fork, got a new one and applied the latest changes. Seems this fixed the issues.

### CODE REVIEW

**Comment by dnovatchev** [comment]
@ChristianGruen 

I think that the following is a better (more precise) formal definition of [**`fn:chain`**](https://qt4cg.org/specifications/xpath-functions-40/Overview.html#func-chain):

```xq
let $chain :=
      (
        let $apply := function($x, $f) 
                       {apply($f, 
                                 if(function-arity($f) eq 1) then [$x]
                                    else if(not($x instance of array(*))) then array{$x}
                                                                          else $x
                                )
                        }
          return
            function($x as item()*, $funs as function(*)*) as item()*
                  {
                    fold-left($funs, $x, $apply)
                  }           
      )                        
```

What do you think?

**Comment by ChristianGruen** [comment]
> I deleted my fork, got a new one and applied the latest changes.

+1 I agree that’s often easier. Looks good now!

> I think that the following is a better (more precise) formal definition of [fn:chain](https://qt4cg.org/specifications/xpath-functions-40/Overview.html#func-chain):

Let me reconsider my suggestions from #733:
* Maybe you could repeat the variable names of the function signature.
* In most code snippet, 2 spaces are used for the indentation.

You could present your code as follows:

```xquery
let $chain := (
  let $apply := function($input, $function) {
    apply($function,
      if(function-arity($function) eq 1) then [ $input ]
      else if(not($x instance of array(*))) then array { $input }
      else $input
    )
  }
  return function($input as item()*, $functions as function(*)*) as item()* {
    fold-left($functions, $input, $apply)
  }           
)
```

Currently:

**Comment by ChristianGruen** [comment]
…and two more typos in the examples that need to be fixed:

```xquery
chain((2, 3), (op("+", $incr(3)))
chain(([1, 2, 3], (count#1, $incr(3)))
```

Some examples have whitespace before opening and closing parentheses, others have not. Maybe this could be unified.

Last but not least, you could reformat the examples to save horizontal space and get rid of the scrollbars, e.g. as follows:

```xquery
chain(
  "abra cadabra",
  (tokenize#2(?, " "), string-join#2(?, "+"))
)

chain((
  chain( ("A    long   message   ", "   long "), 
    (head#1, normalize-space#1, normalize-unicode#1) ),
  chain( ("A    long   message   ", "   long "),  
    (tail#1, normalize-space#1, normalize-unicode#1) )
 ),
 contains#2
)
```

**Comment by dnovatchev** [comment]
Thanks @ChristianGruen, I think your remarks must have been reflected now.

> Last but not least, you could reformat the examples to save horizontal space and get rid of the scrollbars, e.g. as follows:

I am not seeing any horizontal scrollbars - are you using a low-resolution monitor?

These days I am regarding my FHD monitors as almost obsolete, but even on them there is no horizontal scrolling involved when browsing this function's specification.

Thanks,
Dimitre

**Comment by ChristianGruen** [comment]
> Thanks @ChristianGruen, I think your remarks must have been reflected now.

Some of them, yes ;)

> I am not seeing any horizontal scrollbars - are you using a low-resolution monitor?

4K (but in Japan they often have 8K).

image

**Comment by dnovatchev** [comment]
> > Thanks @ChristianGruen, I think your remarks must have been reflected now.
> 
> Some of them, yes ;)
> 
> > I am not seeing any horizontal scrollbars - are you using a low-resolution monitor?
> 
> 4K (but in Japan they often have 8K).
> 
> image

Must be an issue with your browser then. I am using Chrome and even with a resized screen that has less with there are no horizontal scrollbars:

image

Or maybe you have zoomed in with more than 200% (although I still don't see any such effect even with 200% zoom-in.

**Review by ChristianGruen — APPROVED** [review]

**Comment by dnovatchev** [comment]
Great!

So, can now this PR be merged?

And thanks a lot, @ChristianGruen !

**Comment by michaelhkay** [comment]
Could you simplify `if(not($x instance of array(*))) then array{$x} else $x` by removing the `not()` and swapping the then/else branches?

**Comment by dnovatchev** [comment]
@michaelhkay Done.

Should appear shortly as updated html.

File: specifications/xpath-functions-40/src/function-catalog.xml [hunk 1]
```diff
@@ -18280,40 +18296,43 @@ chain( (), true#0)
...
                     <fos:expression>
-                      <eg>
-                        <![CDATA[
+                      <eg><![CDATA[
 chain((1, 2, 3, 4), $product3) 
     ]]></eg>
                     </fos:expression>
-                    <fos:result>Raises a type error ([FOAP0001]) because 4 arguments are supplied to the function but it needs 3 arguments</fos:result>
+                    <fos:error-result error-code="FOAP0001"/>
+                    <!--<fos:result>Raises a type error ([FOAP0001]) because 4 arguments are supplied to the function but it needs 3 arguments</fos:result>-->
```
Comment (ChristianGruen) [inline specifications/xpath-functions-40/src/function-catalog.xml:18324]: Minor suggestion: Obsolete comments should better be removed from PRs.

Comment (dnovatchev) [reply specifications/xpath-functions-40/src/function-catalog.xml:18324]: Done.

**Comment by ndw** [comment]
The group agreed to merge this PR at [meeting 053](https://qt4cg.org/meeting/minutes/2023/11-07.html)

**Pull request merged**